CMakeLists.txt

```diff
@@ -488,134 +509,37 @@ function(
-    add_lit_testsuite(check-picolibc_${variant}
-        "Running picolibc tests: ${variant}"
-        ${CMAKE_CURRENT_BINARY_DIR}/picolibc_tests_${variant}
-        DEPENDS picolibc_${variant} compiler_rt_${variant} llvm-toolchain-runtimes)
-
-    function(picolibc_test test)
-        set(BUILD_PARAMS --config ${variant}_semihost.cfg
-                -T ${CMAKE_CURRENT_FUNCTION_LIST_DIR}/test-support/ldscripts/picolibc_${variant}.ld
-                -nostdlib -lc -lg -lm -lclang_rt.builtins -lsemihost -Oz
-                ${picolibc_SOURCE_DIR}/test/${test}.c ${ARGN} -o ${CMAKE_CURRENT_BINARY_DIR}/picolibc_tests_${variant}/check-picolibc_${variant}_build_${test}.out)
-        string(REPLACE ";" " " BUILD_PARAMS "${BUILD_PARAMS}")
-        set(
-            TEST_COMMAND
-            ${CMAKE_CURRENT_SOURCE_DIR}/test-support/lit-exec-qemu.py
-            --qemu-command ${qemu_command}
-            --qemu-machine ${qemu_machine})
-        if(qemu_cpu)
-            list(APPEND TEST_COMMAND "--qemu-cpu" "${qemu_cpu}")
-        endif()
-        if(qemu_params_list)
-            list(APPEND TEST_COMMAND "--qemu-params=${qemu_params_list}")
-        endif()
-        list(APPEND TEST_COMMAND check-picolibc_${variant}_build_${test}.out)
-        string(REPLACE ";" " " TEST_COMMAND "${TEST_COMMAND}")
-        set(PICOLIBC_LIT_TEST_TEMPLATE ${CMAKE_CURRENT_SOURCE_DIR}/test-support/picolibc-lit-test.template)
-        # Some picolibc tests return non-zero value upon success, then use dedicated test template
-        if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/test-support/picolibc-lit-test-${test}.template)
-            set(PICOLIBC_LIT_TEST_TEMPLATE ${CMAKE_CURRENT_SOURCE_DIR}/test-support/picolibc-lit-test-${test}.template)
-        endif()
-        configure_file(${PICOLIBC_LIT_TEST_TEMPLATE} ${CMAKE_CURRENT_BINARY_DIR}/picolibc_tests_${variant}/${test}.test)
-    endfunction()
+    ExternalProject_Get_Property(picolibc_${variant} BINARY_DIR)
+    configure_file(${CMAKE_CURRENT_FUNCTION_LIST_DIR}/cmake/meson-cross-build.txt.in ${BINARY_DIR}/meson-cross-build.txt @ONLY)
 
-    foreach(test IN LISTS picolibc_tests)
-        picolibc_test(${test})
-    endforeach()
-    # Test disabled due to duplicated symbol name div in test file and in picolibc
-    # picolibc_test(rounding-mode)
-    # Test disabled due to fail on armv8.1m.main_hard_nofp_mve
-    # picolibc_test(try-ilp32)
-    add_dependencies(check-picolibc check-picolibc_${variant})
+    # Building picolibc tests requires compiler_rt to be installed.
+    # Building compiler_rt tests requires picolibc to be installed.
+    # To solve this compiler_rt relies on picolibc to be just installed, not on
+    # the full target, which would include tests. Picolibc tests, are enabled
+    # only in tests step, otherwise, they would be built before install.
+    add_custom_target(compiler_rt_${variant}-available)
+    ExternalProject_Add_StepDependencies(picolibc_${variant} test compiler_rt_${variant}-available)
```

Comment:
Could a standard step be used instead? e.g. `compiler_rt_${variant}-install`

Reply:
It would be more readable, but the issue is, the target has to exist at the point where `ExternalProject_Add_StepDependencies` is called and `compiler_rt_${variant}-install` is created later, by `ExternalProject` call for `compiler_rt`. I added this dummy `compiler_rt_${variant}-available` to resolve the mutual dependency between `picolibc` and `compiler-rt`.

Reply:
> the target has to exist at the point where ExternalProject_Add_StepDependencies is called

Are you sure? Usually in CMake targets don't have to exist at the point the dependency is declared.

Reply:
You are right. CMake happily accepted the `-install` target. I will fix this in next iteration.